### ISSUE TITLE
is there anyone able to explain the data structure of this module?

### ISSUE THREAD

**Issue by sleupold** [ActiveForums/ActiveForums#333]
I just had a look at the database tables of a client site, and don't understand the structure:![af_relations](https://cloud.githubusercontent.com/assets/6377509/10624131/dfb2ca04-7795-11e5-888a-3f39d27cec0e.jpg)
is there anyone able to explain this?

**Comment by sleupold** [comment]
PS: those are all relations within the 31 tables of AF inside this installation.

**Comment by sleupold** [comment]
PPS: for those, who are not familiar with database diagrams. only the solid colored relations are enforced, all others are disabled.

**Comment by supermansrf** [comment]
I'm sure someone will chime in, but it may take a few days.  I'll send out a few emails to see if I can get someone to chime in.  Anything specific you are looking for?

**Comment by sleupold** [comment]
Superman, sure, how do we turn this into a reliable structure?
Clients are complaining about issues arising, as soon as they enable the constraints.

**Comment by sleupold** [comment]
* A ForumsTopic record should for sure not exist without a Topic record existing, this is an obvious bug
* What are the use cases for posts linked to multiple forums?
* Why are replies to this post forum specific?
* how is content connected to Forum Topic and Reply
* I'd expect a simple relational structure Forum > Thread > Post. What is the use case for this construct?
* why do other tables like attachments etc. have no relationship?

**Comment by artfuld** [comment]
Hello @sleupold I just tried to write down the explanation of all this but it's so messed up that it's hard to follow.

Basically a topic contains the ContentId for the first post and then the reply-table contain the contentId for all the replies.

I'm not sure if Attachments or Content_Attachments is the one in use, but I'm 33% sure that one is in use and the other is old code.

Same with the tables Permissions and Security, one of them is from the old code,

As there's more or less no one still here from the time this was a commercial product and the owner of the project has changed more than once, I'm not sure if there's anyone still here that can answer as to who did the changes and why the changes was made.

I have had hopes about first change to dal2 but the way the template works this is a far to big thing and it probably would be faster to create a new forum from scratch and steal some of the helpers from this forum.

**Comment by PaulCam** [comment]
I wonder if Will Morgenweck would help out here. He was the original developer of the module I think.

**Comment by sleupold** [comment]
Artfuld, thanks for your reply, but it doesn't answer the question about ForumTopics table - is there a use case, where a single discussion may be opened in multiple forums and be answered in each individually?

**Comment by artfuld** [comment]
No Sebastian, there's no such function today where you can do that. You can't even have a post show up in more than one forum.

**Comment by sleupold** [comment]
Joakim, thanks, that meets my experience using AF, but the data structure is built just for this scenario - very strange.

**Comment by supermansrf** [comment]
@sleupold there are MANY things that need looked at within this module.  When Active Modules went to work for DNN and the module was prepped for release to the community, there was code removed that was under commercial license, along with many unfinished projects that just went by the wayside.  Then others took over and tweaked things and what not.  So far everything that i've seen has been enhancement or break-fix.  

Truth be told this module has never gotten the attention that I have felt it deserves.  Anything you wish to contribute or help make better I'm completely for.  I have always felt like there is a huge hole for a great forum in DNN.  I also had discussed making a new forum from scratch with my primary developer, but truth be told I just couldn't justify the expense for a few sites.  

At one point in time I had been in discussions with someone who was going to work on making the database more efficient, but as you are aware (I'm sure) getting volunteers isn't always easy.  

Anything you see that you would like to correct, feel free.  Your reputation proceeds itself.  At least for me.  I know your abilities.

**Comment by sleupold** [comment]
Seth, thanks for your reply. I will need some time to come up with database improvements, maybe starting by adding relations and optimizing indexes.

**Comment by supermansrf** [comment]
Rome wasn't built in a day lol.  I've been waiting for years so anything less than before I'm 6 feet under is fine with me.  I just appreciate you offering to help.
